From 9c47dd8c7f7e3026cefebf18659fbe551912b2ae Mon Sep 17 00:00:00 2001
From: Venkat Gopalakrishnan <venkatg@codeaurora.org>
Date: Fri, 19 Apr 2013 10:27:36 -0700
Subject: [PATCH] msm: sdio: Fix memory leak

Release the tty_driver struct on all exit paths to avoid memory leaks.

Change-Id: Ib4529f120e70b343e55e72122f19a877a8bc0503
Signed-off-by: Venkat Gopalakrishnan <venkatg@codeaurora.org>
---
 arch/arm/mach-msm/sdio_al_dloader.c |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/arch/arm/mach-msm/sdio_al_dloader.c b/arch/arm/mach-msm/sdio_al_dloader.c
index a9bdf01..f3effa8 100644
--- a/arch/arm/mach-msm/sdio_al_dloader.c
+++ b/arch/arm/mach-msm/sdio_al_dloader.c
@@ -2487,6 +2487,7 @@ int sdio_downloader_setup(struct mmc_card *card,
 		pr_err(MODULE_NAME ": %s - tty_register_device() "
 			"failed\n", __func__);
 		tty_unregister_driver(sdio_dld->tty_drv);
+		put_tty_driver(sdio_dld->tty_drv);
 		kfree(sdio_dld);
 		return PTR_ERR(tty_dev);
 	}
@@ -2536,6 +2537,7 @@ exit_err:
 	if (result)
 		pr_err(MODULE_NAME ": %s - tty_unregister_driver() "
 		       "failed. result=%d\n", __func__, -result);
+	put_tty_driver(sdio_dld->tty_drv);
 	kfree(sdio_dld);
 	atomic_set(&sdio_dld_setup_done, 0);
 	return status;
@@ -2554,12 +2556,12 @@ static void sdio_dld_tear_down(struct work_struct *work)
 	tty_unregister_device(sdio_dld->tty_drv, 0);
 
 	status = tty_unregister_driver(sdio_dld->tty_drv);
-
 	if (status) {
 		pr_err(MODULE_NAME ": %s - tty_unregister_driver() failed\n",
 		       __func__);
 	}
 
+	put_tty_driver(sdio_dld->tty_drv);
 	kfree(sdio_dld);
 	atomic_set(&sdio_dld_in_use, 0);
 	atomic_set(&sdio_dld_setup_done, 0);
-- 
1.7.8.3

